### DISCUSSION TITLE
Is it possible to calculate default values dynamically?

### DISCUSSION THREAD

**Discussion by pfmoore** [copier-org/copier#672] [Q&A] [answered]
I want to prompt the user for their name, email and github username, but I'd like to provide sensible defaults (which can be obtained by running `git config` to get the appropriate values). Is it possible to create defaults dynamically like this?

**Comment by pawamoy** [comment]
I suppose Copier could retrieve the values itself and inject them in the Jinja context used for questions and answers. We could then set defaults with something like `default: "{{ _git_user_name }}"` or `default: "{{ _git_user_email }}"`.

**Reply by pfmoore** [reply]
I guess so, but who would decide what was worth including? I was hoping there would be a way to execute code to generate defaults, as that's a bit more flexible (although it's also a security risk, so I'd be fine if the ability was limited somehow).

Is there a hook that templates can use to adjust the Jinja context? Or is this something a Jinja plugin could do? (I don't know much about Jinja plugins).

**Reply by pawamoy** [reply]
I have successfully written a Copier/Jinja extension that can be used to alter the *rendering* context (now in the org: https://github.com/copier-org/copier-templates-extensions). I'm not sure it would be possible to do the same thing for the *prompt* context. I'll need to check Copier's code flow 🙂

**Reply by pfmoore** [reply]
I just tested the extension I wrote (see the answer I added below) and that seems to work fine in the prompt context. Thanks for the comment, I wasn't even aware this could be an issue 🙂

**Comment by pfmoore** [comment]
Actually, having put the question that way, I realised that this is at least partially covered in the FAQ [here](https://copier.readthedocs.io/en/latest/faq/#how-can-i-alter-the-context-before-rendering-the-project). And in fact, writing a simple Jinja extension that just adds stuff to the environment isn't particularly hard:

```python
from jinja2.ext import Extension


class TestExtension(Extension):
    def __init__(self, environment):
        super().__init__(environment)

        # add the defaults to the environment
        environment.globals["some_variable_name"] = "The value you want"
```

So I guess the answer is that to do this I can write a Jinja extension to do this, and load that extension in `copier.yml`.

Have I missed any subtleties here? As I said, I don't know much about Jinja plugins (the docs say "as long as no template was loaded it’s safe to modify this" - it seems to work here so I assume extensions are initialised early enough that we're OK).

It's a bit annoying to have to depend on an extension being installed, but the `copier_templates_extensions` approach seems to end up with the extension code being included in the generated project, which isn't ideal either. So I'll probably stick with a plain extension.

**Comment by yajo** [comment, marked as answer]
Have you tried with [the shell extension](https://pypi.org/project/jinja2-shell-extension/)?

It should let you do something like:

```yaml
# copier.yaml
_jinja_extensions:
  - jinja2_shell_extension.ShellExtension
email:
  default: "{{ 'git config user.email'|shell }}"
```

**Reply by pfmoore** [reply]
Ooh, nice! I didn't know that existed. I did wonder if this was something that might already exist, but I couldn't find a good way of looking for it.

**Reply by yajo** [reply]
You have [a hint in the docs](https://copier.readthedocs.io/en/stable/configuring/#jinja_extensions) (I used that one FWIW 😆).

**Reply by pfmoore** [reply]
Bah, I missed that. Sorry, I've just discovered copier in the process of trying to find a replacement for my current cookiecutter templates, and I have a bad habit of getting over-enthusiastic and jumping straight into complex stuff, without taking time to get properly familiar with the basics. Sorry about that and thanks for your patience and help 🙂

**Reply by yajo** [reply]
😄 don't worry, we're full of cookiecutter refugees here.

This might help you: https://copier.readthedocs.io/en/stable/comparisons/

Probably it's not complete, but PRs are always welcome!
